충남대 BE_윤정훈 3주차 과제 (1단계, 2단계)

yunjunghun0116 opened the pull request:
## 1단계 및 2단계 과제 수행 과정

1. README.md 작성
2. 일반 클래스로 만들어진 도메인 계층의 클래스를 엔티티 클래스로 작성
3. JdbcTemplate 기반으로 작성된 레포지토리 계층을 JPA 레포지토리로 전환 작업
4. 여러가지 시나리오에 대해서 테스트코드 작성


## 오늘 여쭤보려 했던, 과제를 수행하는동안 고민했던 내용입니다!

과제를 수행하는 과정에서 
```
public WishProduct findWishProductWithId(Long id) {
        var wishProduct = wishProductRepository.findById(id);
        if (wishProduct.isEmpty()) throw new NotFoundElementException("존재하지 않는 리소스에 대한 접근입니다.");
        return wishProduct.get();
    }
```
와 같이 id를 바탕으로 Product, WishProduct, Member 등을 조회하는 기능이 반복되는 부분이 보여 이 부분을 각 Service마다 
findProductById, findMemberById 와 같은 함수를 만들고 이를 조회하는 부분에서 
```
private final MemberService memberService;
...
var member = memberService.findMemberWithId(memberId);
```
와 같이 Service를 주입받고, 참조하도록 하였는데 추후 다양한 도메인이 추가됨에 따라 다른 Service 계층에 대한 의존성이 더욱 추가되고, 순환참조가 발생할 수 있겠다는 생각에 조회만을 수행하는 Helper클래스를 생성하였습니다.

해당 클래스는 
```
@Component
public class RepositoryReader {
    ...
    public <T> T findEntityById(JpaRepository<T, Long> repository, Long id) {
        var entity = repository.findById(id);
        if (entity.isEmpty()) throw new NotFoundElementException("존재하지 않는 리소스에 대한 접근입니다.");
        return entity.get();
    }
}
``` 
위와 같이 제네릭 메서드로 구성되어있고, 이를 외부에서는 주입받고 findEntityById(repository, id)를 통해 객체를 조회할 수 있도록 하였습니다.
위 경우 순환참조의 위험성은 최소화하고, 도메인이 추가됨에 따라 Service 계층의 의존성이 추가되지 않아 목적은 충족하지만 결국 Helper클래스에 대한 의존성이 추가되게 됩니다.
그래서 이에 대해서 어떻게 해결할까? 고민을 하다가 JpaRepository도 제네릭으로 구성되어 있다는 것을 알게 되었습니다.
그래서 다시 한번 위 코드를
```
@NoRepositoryBean
public interface BaseRepository<T, ID> extends JpaRepository<T, ID> {
    default T findByIdOrThrow(ID id) {
        return findById(id)
                .orElseThrow(() -> new NotFoundElementException("존재하지 않는 리소스에 대한 접근입니다."));
    }
}
```
와 같이 findByIdOrThrow 기능을 수행하는 함수를 만들고 다른 Repository는
```
public interface MemberRepository extends BaseRepository<Member, Long>
```
처럼 이를 상속받도록 하여 구현을 했습니다.

위 과정에서 궁금한 점은 현업에서 위와 같이 조회와 관련하여 중복되는 코드가 지속적으로 발생하는 경우
1. Helper(혹은 Util)클래스를 만들고 빈으로 등록하여 의존성 주입받아 사용한다.
- 장점 : 조회 후 예외 처리에 대한 코드 중복을 해소할 수 있고, Service 계층간 의존하지 않아도 된다. 또한 다양한 기능을 추가하는데 있어 용이하다.
- 단점 : 필요한 위치에 의존성을 주입받아 사용해야 한다. 

2. BaseRepository를 만들고, 이를 상속받도록 구현한다.
- 장점 : 위와 같이 코드 중복을 해소하고, Service 계층간 의존하지 않아도 된다. 또한 추가적인 의존성의 주입이 필요없이 Service계층에서는 해당 Repository만 주입받으면 사용할 수 있다.
- 단점 : 모든 Repository가 BaseRepository를 상속받는 구조가 강제된다. 또한 3~4개 등 일부 Repository에서만 중복되는 코드의 경우 상속을 한번 더 받아 사용해야 하는 등 기능을 추가하는데에 있어 번거로움이 발생할 수 있다.

과제를 진행하면서 중복에 대해서 문제를 해결하기 위해 1번 선택지를 통해 개선을 했고, 메서드 호출시마다 헬퍼 클래스를 주입받아야 한다는 번거로움 때문에 해결방법을 계속 고민하다 2번 선택지를 바탕으로 번거로움을 해결했습니다.

다만 실제 서비스에서는 훨씬 많은 도메인이 존재하고, 그에 따라 더 복잡한 로직이 존재한다고 생각합니다.
일부 3~4개 와 같은 Repository에서만 중복되는 코드의 경우 BaseRepository를 다시 한번 상속받아야 하는 등 중복되는 기능이 추가될수록 상속 계층 구조가 복잡해질 수 있다고 생각합니다.

그래서 2번 선택지가 현재로서는 최선이지만 미래를 생각하면 최선이 아닐 수도 있겠다는 생각입니다.

멘토님은 위와 같은 경우 1번 선택지와 2번 선택지, 혹은 다른 선택지를 통해 어떤식으로 중복코드와 의존성을 해결하실지 궁금합니다.

## 궁금한 부분

Product와 WishProduct 가 서로 연관관계를 맺고 있는데 WishProduct에서
```
public void addProduct(Product product){
 product.getWishes().add(this);
 this.product = product;
}
```
와 같이 서로간의 관계 설정을 맡고있고, WishProductService의 saveWishProduct 메서드에서는
```
private WishProduct saveWishProductWithWishProductRequest(Product product, Member member, Integer count){
var wishProduct = new WishProduct(count);
wishProduct.addProduct(product);
wishProduct.addMember(member);
var savedProduct = wishProductRepository.save(wishProduct);
return savedProduct;
}
```
를 수행하는데, 양방향 연관관계에 대해 객체를 생성하여 저장하는 과정에 대해서 아직 이해가 조금 부족한것 같아 이부분에 대해서 로직을 한번 더 자세히 확인 부탁드립니다!

yunjunghun0116 committed:
feat: Product 엔티티 매핑 작업 수행

Product 엔티티 매핑 작업 수행

yunjunghun0116 committed:
feat: ProductOption 엔티티 매핑 작업 수행

ProductOption 엔티티 매핑 작업 수행

yunjunghun0116 committed:
feat: Member 엔티티 매핑 작업 수행

Member 엔티티 매핑 작업 수행

yunjunghun0116 committed:
feat: Member 엔티티 매핑 작업 수행

Member 엔티티 매핑 작업 수행

yunjunghun0116 committed:
test: 기존 schema.sql -> BeforeEach, AfterEach 를 통해 기본 데이터를 추가해준다.

기본 데이터의 추가를 함수 형태로 추가해준다.

yunjunghun0116 committed:
test: 연관관계 매핑을 통해 부모객체에서 자식객체를 조회할 수 있다.

연관관계 매핑을 통해 부모 객체에서 자식 객체를 조회할 수 있는 테스트를 추가한다.

yunjunghun0116 committed:
remove: cascade와 orphanremoval 을 사용함에 따라 외래키제약조건위배예외 제거

외래키제약조건위배예외 제거

yunjunghun0116 committed:
add: hibernate.ddl-auto = update로 주어 엔티티 매핑에 따라 스키마를 업데이트하도록 변경

기존 : 종료시 삭제했는데, 이경우 drop table이 수행되기에 update로 변경

yunjunghun0116 committed:
test: WishProductServiceTest에 BDD 적용

테스트 BDD 적용

yunjunghun0116 committed:
test: ProductServiceTest에 BDD 적용

BDD 적용

yunjunghun0116 committed:
refactor: create로 분리한 함수에서 save를 수행하고, 그에 대한 결과만을 반환받아 작업하도록 save 함수 로직 변경

save함수 로직의 일부 이동 작업 수행

yunjunghun0116 committed:
test: Service계층 테스트 BDD 적용

서비스계층 테스트 BDD 적용

yunjunghun0116 committed:
test: 컨트롤러 계층의 테스트 BDD 적용

시나리오 기반 테스트 룰 적용

yunjunghun0116 committed:
feat: Repository에서 id를 통해 조회후 예외에 대한 처리 작업을 수행하는 RepositoryReader 컴포넌트 추가

컴포넌트 추가후 이를 통해 작업 수행

yunjunghun0116 committed:
test: 테스트 값 수정

테스트 값 수정

yunjunghun0116 committed:
refactor: ReaderClass 로 작성된 코드를 BaseRepository를 만듦으로써 의존성 줄임

의존성을 줄이기 위해 BaseRepository를 만듦

yunjunghun0116 committed:
remove: 이미 구현되어있는 함수에 대해서 제거작업수행

제거 작업 수행

yunjunghun0116 committed:
refactor: id와 생성 및 수정시간을 갖는BaseEntity를 상속받도록 하여 공통 부분을 수정한다.

BaseEntity를 상속받도록 하여 공통 부분을 수정한다.

yunjunghun0116 committed:
remove: 사용되지 않는 필드의 제거

사용되지 않는 필드의 제거

yunjunghun0116 committed:
docs: README.md 2단계 기능 요구사항 작성

2단계 기능 요구사항 작성

yunjunghun0116 committed:
docs: README.md 수정

README.md 수정

yunjunghun0116 committed:
docs: README.md 수정

asebn1 commented on src/main/java/gift/dto/ProductOptionResponse.java:
```diff
@@ -1,7 +1,7 @@
-public record ProductOptionResponse(Long id, Long productId, String name, Integer additionalPrice) {
-    public static ProductOptionResponse of(Long id, Long productId, String name, Integer additionalPrice) {
-        return new ProductOptionResponse(id, productId, name, additionalPrice);
+public record ProductOptionResponse(Long id, ProductResponse productResponse, String name, Integer additionalPrice) {
+    public static ProductOptionResponse of(Long id, ProductResponse productResponse, String name, Integer additionalPrice) {
+        return new ProductOptionResponse(id, productResponse, name, additionalPrice);
     }
```
`ProductResponse`를 반환하는게 좋은지 다시 고민해봅시다!
만약 `ProductResponse`의 필드값 변경이 있다면 `ProductOptionResponse`에도 영향이 있을 것 같네요

yunjunghun0116 replied:
ProductResponse를 반환하는 상황과 사용되는 상황에 대해서 다시 한번 고민해보았습니다.

우선 제 생각은 WishProduct, ProductOption 에서는 `상품의 모든 정보가 필요하지 않을 수 있다` 입니다.. 예를 들면 상품의 상세 설명, 상품에 대한 이미지 등 모든 정보가 필요하지는 않을 수 있다고 생각합니다.

하지만 Product에 대한 조회를 수행하는 이용자에게는 모든 정보가 필요할 수 있기에 ProductResponse는 그 자체로 필요하고, 그러면 이를 해결하기 위해 새로운 DTO가 필요하다는 결론이 나왔습니다.

WishProduct, ProductOption 과 같은 상품 관련 객체(엔티티)에서는 필요로 하는 상품정보는 모든 정보가 아닌 상품의 "기본 정보" 혹은 "필수 정보", 예를 들면 상품의 이름과 가격, 로직을 수행하는데 필요한 ID 등이라고 생각했습니다.

그렇기에 상품의 기본 정보만을 담는 `ProductBasicInformation` 이라는 DTO를 만들고, 이 DTO는 WishProduct, ProductOption 과 같이 상품의 기본정보만을 필요로 하는 ResponseDTO 에 기존 ProductResponse 대신 ProductBasicInformation 을 담아줌으로써 클라이언트에게는 오직 "클라이언트가 필요로 하는 정보" 만을 제공할 수 있도록 DTO를 추가하였습니다.

위 과정을 통해 Product와 ProductResponse의 필드값이 변경되는 경우가 있다 하더라도 기본정보가 변경되지 않는한 WishProductResponse, ProductOptionResponse 에는 영향이 없도록 하여 객체간 영향도도 기존보다 낮게 개선하였습니다.

asebn1 commented on src/main/java/gift/model/Member.java:
```diff
@@ -1,22 +1,39 @@
+    @Enumerated
+    @Column(name = "role")
     private MemberRole role;
+    @OneToMany(mappedBy = "member", cascade = CascadeType.ALL, orphanRemoval = true)
+    private List<WishProduct> wishes = new ArrayList<>();
```
cascade를 ALL로 설정해주었네요
다른 설정값엔 무엇이있을까요? ALL을 적용해야할까요?

yunjunghun0116 replied:
다른 설정값에는
CascadeType.PERSIST, MERGE, REMOVE, DETACH, REFRESH가 있습니다.

위 속성값은 모두 영속성과 관련이 있고, 영속성(컨텍스트) 는 DB에 존재하는 데이터를 객체형태로 관리하는 저장소입니다.
+ PERSIST 는 A와 연관되어있는 엔티티 B, C가 있고, A에 대해서 save를 수행하게 되면 B, C도 함께 save되어 영속 상태가 되도록 하는것
+ MERGE 는 연관된 엔티티가 함께 merge(update) 되도록 하는것
+ REMOVE 는 연관된 엔티티가 함께 삭제되도록 하는것이며, 이 경우 고아객체가 생길 수 있기에 orphanremoval 를 통해 발생한 고아객체에 대한 처리까지 수행할 수 있음
+ DETACH 는 영속성 컨텍스트에서 관리를 하지 않겠다고 선언하는것이며, 연관된 엔티티 모두 관리를 하지 않게되는것
+ REFRESH 는 영속성 컨텍스트에서 DB에 접근해 다시 불러오는 작업을 연관된 엔티티 모두 수행하는것

또한 find, update 등의 작업을 service계층에서 하나의 트랜잭션으로 묶어 작업을 수행하기 때문에 ALL을 적용하지 않고, PERSIST를 통해 연관된 엔티티에 대해서 함께 영속 상태가 될 수 있도록 하여 적절한 CascadeType을 적용할 수 있겠네요!

그리고, DETACH는 오히려 JPA를 사용하는 입장에서 EntityManager의 관리 범위 밖으로 넘어가는 것이기에 위험할수 있다고 생각했습니다!

그렇기에 저는 PERSIST, MERGE, REMOVE 를 통해 연관된 객체에 있어 추가, 수정, 삭제에 관해서만 영속성 관리를 하도록 
`{CascadeType.PERSIST,CascadeType.MERGE,CascadeType.REMOVE}` 로 필요한 설정만을 해주었습니다.

yunjunghun0116 replied:
추가로 로직 복잡도를 개선할 수 있다고 생각하여 우선 다대일 단방향 관계로 설정하여 연관관계를 맺어 주고, 현재는 반대방향으로 탐색하는 기능이 필요로 하지 않기에 ManyToOne만을 사용했습니다.

로직 복잡도에 대한 생각
1. 자식 엔티티를 추가할 경우 부모 엔티티도 조회하여 추가해주는 작업을 수행해야합니다.
2. 자식 엔티티를 삭제할 경우 부모 엔티티를 조회하여 `member.getWishes.remove(this)` 로직도 수행하여 연관관계를 끊어준 후에 삭제를 해야합니다.

와 같은 이유에서 이 부분을 다대일 단방향 관계로 설정하더라도 필요한 경우 자식Repository에서 findAllByProductId 등을 통해 필요한 데이터를 가져올 수 있다고 생각했습니다.

이 경우 부모 객체를 삭제 시 외래키 제약조건에 위배되어 삭제를 할 수 없게 되는데, 이를 해결하기 위해 repository에서 외래키를 바탕으로 삭제하는 deleteWishProductsByMemberId, ProductId 와 같은 메서드를 생성하여 해결했습니다.

asebn1 commented on src/main/java/gift/model/Product.java:
```diff
@@ -1,20 +1,36 @@
+    @Column(name = "price")
     private Integer price;
+    @NotNull
+    @Column(name = "image_url")
     private String imageUrl;
+    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
+    private List<ProductOption> options = new ArrayList<>();
+    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
+    private List<WishProduct> wishes = new ArrayList<>();
```
product가 가지는 옵션들로 `List<ProductOption>`가 있는게 직관적으로 이해가 가네요
- `wishes`를 product에서 관리해야할까요?
- 네이밍은 productOptions은 어떨까요?

yunjunghun0116 replied:
+ wishes는 Product에서는 관리할 필요 없습니다!
이유는 논리적 연관관계가 현재는 발생하지 않고, 그로 인해 역방향(Product -> WishProduct) 조회 역시 로직상 발생하지 않기 때문입니다.
그래서 논리적 연관관계가 발생하는 wishes는 Member와 양방향 연관관계를 맺고, wishes는 Member에서만 관리하도록 하였습니다.

+ productOptions가 조금 더 직관적인 것 같아 네이밍 변경하였습니다!

asebn1 commented on src/main/java/gift/model/WishProduct.java:
```diff
@@ -1,35 +1,43 @@
-    private Long id;
-    private Long productId;
-    private Long memberId;
+@Entity
+@Table(name = "wish_product")
+public class WishProduct extends BaseEntity {
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "product_id", referencedColumnName = "id")
+    private Product product;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "member_id", referencedColumnName = "id")
+    private Member member;
+    @NotNull
+    @Column(name = "count")
```
`Product`와 `Member`의 다대다 관계를 위해 중간 테이블로 `WishProduct`를 만들어주셨군요~ 두 속성도 `@NotNull`이 들어가면 좋을 것 같아요

yunjunghun0116 replied:
멘토님 말씀대로 `WishProduct` 는 `누가(Member)`, `무엇을(Product)` 에 대한 정보는 수정되지 않을것이라 판단하여 product와 member를 생성자를 통해 초기에 받도록 하였고 두 속성에 대해 `@NotNull` 어노테이션을 붙여주어 서비스, DB 모두 null값에 대한 제약을 걸어주었습니다!

asebn1 commented on src/main/java/gift/repository/BaseRepository.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.repository.NoRepositoryBean;
+
+@NoRepositoryBean
+public interface BaseRepository<T, ID> extends JpaRepository<T, ID> {
+    default T findByIdOrThrow(ID id) {
+        return findById(id)
+                .orElseThrow(() -> new NotFoundElementException("존재하지 않는 리소스에 대한 접근입니다."));
+    }
+}
```
지금 구조도 괜찮다고 생각합니다~
다만 추가로 생각해볼 내용이 있습니다

1. 하나로 묶은 에러얼럿이 `존재하지 않는 리소스에 대한 접근입니다`인데, 어느 엔티티에서 발생한 에러인지 추적하기 어려울 것 같아요
2. 조회 시 반환값에서 Optional 반환을 맞추어주세요. Member의 경우 `Optional<Member> findByEmail(String email);`로 사용하고있네요
3. `.orElseThrow(() -> new NotFoundElementException("존재하지 않는 리소스에 대한 접근입니다."));`의 에러처리는 서비스/레포지토리 중 어디에서 하는게 좋다고 생각하시나요?
4. 서비스가 계속 커지는 상태에서 항상 모든 레포지토리가 `findByIdOrThrow`를 사용한다는 보장을 할 수 있을까요?

yunjunghun0116 replied:
멘토님 말씀처럼 위 코드로 개발을 하게 될 경우 `어느 상황에서`, `어떤 엔티티가` 예외가 발생했는지에 대해서 추적하기 어려울 것 같습니다.
그래서 저는 기존코드가 오히려 하이리스크 로우리턴 코드가 될 수 있다고 생각하였습니다.
또한 서비스계층에서 같은 findById 로직을 수행하는것처럼 보일 수 있지만 `에러얼럿이 달라질 수 있고, 결국 중복 코드가 아니라 다른 코드다!` 라는 판단을 하고 이 부분은 서비스계층에서 필요한 경우 findById를 수행하도록 하였습니다.

또한 4번 질문처럼 findByIdOrThrow가 모든 레포지토리에서 사용한다고 보장할 수 없습니다. 그럴 경우 오히려 불필요한 상속 구조가 만들어진 것이기에 이 부분도 서비스계층에서 필요한 경우 findById를 수행하도록 하여 개선하였습니다.

2번 질문에 대해서 Optional<엔티티>로 반환하도록 하여 존재하지 않는 객체에 대해서도 Optional.empty()를 반환하도록 하여 존재하지 않는 객체에 대해 명시적으로 처리해주기 위해서 MemberRepository의 findByEmail, WishProductRepository의 findbyProductAndMember 등을 모두 Optional로 반환하도록 하였습니다.

3번 질문에 대해서는 너무 명확하게도, 에러 처리는 서비스 계층에서 하는게 좋다고 생각합니다.
그 이유는 id를 통해 DB에 접근했을때 응답이 존재하지 않더라도, Repository 계층에서는 정상적으로 동작한것이고, Service계층에서 로직상 발생한 에러이기 때문입니다!

asebn1 commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -1,10 +1,11 @@
+import org.springframework.transaction.annotation.Transactional;
 
 @Service
+@Transactional
```
`@Transactional`을 사용해주셨네요 👍 
어떤 장점이 있을까요?

yunjunghun0116 replied:
우선, 트랜잭션은 하나의 메서드 내에서 다양한 Repository에 접근하여 조작하는 과정을 `하나의 논리적 단위` 로 묶어주는 역할을 합니다.
또한 영속성 컨텍스트와 생명주기를 동일시하기에 이미 조회한 정보를 다시 조회하는 경우 영속성 컨텍스트 내부의 `1차 캐시` 에서 꺼내오기에 속도가 빠를 수 있습니다.
그리고 전체 로직중 일부의 로직만 잘못된 경우에 필요한 경우 하나의 프로세스의 원자성을 보장하기 위해 기존 상태로 되돌아가야 하는 기능이 필요할 수 있는데 이러한 경우 트랜잭션의 롤백을 사용해 돌아갈 수 있습니다.

이처럼 하나의 메소드 처리 과정(프로세스) 의 원자성을 보장할 수 있고 영속성 컨텍스트와 결합되어 성능 향상도 기대할수 있다는 장점이 있습니다.

asebn1 commented on src/main/java/gift/service/ProductOptionService.java:
```diff
@@ -34,30 +41,38 @@ public ProductOptionResponse getOption(Long id) {
     }
 
     public List<ProductOptionResponse> getOptions(Long productId) {
-        return optionRepository.findAll(productId)
+        return optionRepository.findAllByProductId(productId)
                 .stream()
                 .map(this::getProductOptionResponseFromProductOption)
                 .toList();
     }
```
`@Transactional`에서 `readOnly`를 고민해볼까요? 
적용하면 왜 성능이 향상될까요?

yunjunghun0116 replied:
트랜잭션은 데이터베이스의 상태를 일관성 있게 유지하기 위해서 데이터베이스에 존재하는 데이터에 접근할 때 다른 트랜잭션은 접근하여 조작할 수 없도록 락을 걸게 됩니다.

하지만 readOnly=true로 설정하게 되면 락을 설정하지 않아 다른 트랜잭션들과의 동시성 충돌 가능성이 줄어듭니다.
그리고 엔티티의 변경 감지를 위한 스냅샷을 만들지 않아도 되기에 성능 상 이점이 있습니다.

추가적으로 Insert, Update, Delete 가 진행되지 않는다는 것을 명시적으로 알릴 수 있어 가독성이 향상될 수 있습니다.

asebn1 reviewed (changes requested):
정훈님 안녕하세요~
이번에 처음 JPA를 적용해주셨는데 많이 고민하신 흔적이 보이네요 👍 
고민하면 좋을 부분을 코멘트 드렸습니다. 

---

> 멘토님은 위와 같은 경우 1번 선택지와 2번 선택지, 혹은 다른 선택지를 통해 어떤식으로 중복코드와 의존성을 해결하실지 궁금합니다.

말씀 주신 1번 같은 경우는 헬퍼를 통해 독립적으로 사용할 수 있는데 이게 장점이자 단점이 아닐까 싶어요. 팀 프로젝트라면 헷갈릴만한 포인트가 될 수 있지 않을까 싶습니다
`Optional<Member> findById(Long id);` 한 줄 추가되는 거라 개인적으로 큰 부담으로 느껴지진 않았습니다. 제안주신 2번도 팀 내 합의가 된다면 코드중복을 줄여주어 나쁘지 않다고 봐요.
다만 리팩토링이 될 수 있다는 점을 기억해주세요!! (물론 지금은 많이 고민해보고 시도해보는게 좋습니다)

yunjunghun0116 committed:
refactor: repository의 상속 구조 개선을 위한 BaseRepository 제거

BaseRepository를 제거

yunjunghun0116 committed:
remove: 불필요한 연관관계는 지우는 작업 추가

불필요한 연관관계 - 위시 리스트와 상품 사이 양방향 연관관계 -> 단방향 연관관계로 설정

yunjunghun0116 committed:
feat: 외래키제약조건에 위배되지 않도록 자식엔티티가 존재할경우 해당 엔티티부터삭제 후 부모엔티티를 삭제하도록 기능 추가

삭제이 외래키제약조건 고려해서 삭제하도록 기능 추가

yunjunghun0116 committed:
add: NotNull 제약조건 각각의 연관관계 객체에 대해서도 추가

NotNull 제약조건 추가

yunjunghun0116 committed:
add: WishProductResponse, ProductOptionResponse 와 같은 DTO에서 상품정보를 담기 위한 ProductBasicInformation DTO를 담도록 추가

ProductBasicInformation DTO 추가

yunjunghun0116 committed:
refactor: 양방향 연관관계에 있어 불필요한 CascadeType은 제외하고, 필요한 설정만을 하도록 변경

필요한 설정만을 하도록 변경

yunjunghun0116 committed:
add: Member엔티티의 email 속성에 unique 속성 추가

unique 속성 추가

yunjunghun0116 committed:
refactor: 조회 기능에 대해서 readonly 속성을 부여

readonly 속성을 부여해줌

yunjunghun0116 committed:
refactor: 다대일 양방향 관계 대신 단방향 관계로 변경

이유 : delete의 경우 cascade를 할 경우 N+1 문제가 발생하게 되기에 우선 단방향 관계로 변경하고, 추후 이부분에 대해서 JPQL로 쿼리를 변경할 예정
그러면 기존에 있던 역방향 조회는? 이미 해당 기능을 구현하는 service 계층에서의 조회기능이 존재하기에 이부분을 사용하여 조회가 가능하도록 함

asebn1 reviewed (approved):
2단계도 잘 마무리해주셨네요!
고생하셨습니다 👏
2단계는 이만 머지할게요~
3단계도 화이팅!!
